e2e: test canary updates for system jobs

pkazmierczak opened the pull request:
Internal ref: https://hashicorp.atlassian.net/browse/NMD-891

pkazmierczak commented:
```
-: no non-test Go files in /home/runner/work/nomad/nomad/e2e/scheduler_system
hclogvet: analysis skipped due to errors in package
hclogvet: analysis skipped due to errors in package
hclogvet: analysis skipped due to errors in package
```
ehm, yeah, I'm not sure what that's all about. I'll figure this out, sorry.

tgross commented on e2e/e2e_test.go:
```diff
@@ -25,7 +25,6 @@ import (
 	_ "github.com/hashicorp/nomad/e2e/quotas"
 	_ "github.com/hashicorp/nomad/e2e/scalingpolicies"
 	_ "github.com/hashicorp/nomad/e2e/scheduler_sysbatch"
-	_ "github.com/hashicorp/nomad/e2e/scheduler_system"
```
Note the comment in the group below: we import all the tests anyways so that we get a compile check on commit.

jrasell commented on e2e/scheduler_system/systemsched_test.go:
```diff
@@ -0,0 +1,198 @@
+func testCanaryUpdate(t *testing.T) {
+	_, cleanup := jobs3.Submit(t,
+		"./input/system_canary_v0.nomad.hcl",
+		jobs3.WaitComplete("group"),
```
This is waiting for a task group named "group" which doesn't exist in the job spec.

tgross commented on e2e/scheduler_system/systemsched_test.go:
```diff
@@ -0,0 +1,198 @@
+	for _, a := range allocs {
+		if a.JobVersion == 1 {
+			must.True(t, a.DeploymentStatus.Canary)
+			canaryAllocs = append(canaryAllocs, a)
```
We don't use this slice other than to get its length, so we could track counts if we wanted instead.

jrasell commented on e2e/scheduler_system/systemsched_test.go:
```diff
@@ -0,0 +1,198 @@
+	deploymentsApi := job2.DeploymentsApi()
+
+	// wait for the canary allocations to become healthy
+	timeout := time.After(30 * time.Second)
```
This timeout is used for multiple loops rather than just the canary allocations, could we move it above this comment and note that if this is expected?

jrasell commented on e2e/scheduler_system/systemsched_test.go:
```diff
@@ -0,0 +1,198 @@
+	// promote canaries
+	deployments, _, err := deploymentsApi.List(nil)
+	must.NoError(t, err)
+	must.SliceLen(t, 2, deployments)
```
This check uses a hardcoded two, does that mean it would fail if we change the number of clients available to the cluster or some fail to deploy properly? If so, I wonder if we can use an API call to determine how many we expect?

Same comment for line 157 and 197.

jrasell reviewed (commented):
If the two job specification are identical except for the env var, we could possibly use `jobs3.ReplaceInJobSpec` to be able to use a single spec and alter it between deployments. A few inline questions too.

pkazmierczak commented on e2e/scheduler_system/systemsched_test.go:
```diff
@@ -0,0 +1,198 @@
+func testJobUpdateOnIneligbleNode(t *testing.T) {
+	job, cleanup := jobs3.Submit(t,
+		"./input/system_job0.nomad",
+		jobs3.WaitComplete("group"),
```
oof copy pasta

pkazmierczak commented:
@tgross @jrasell I addressed all of your comments (I think) in 01921e6. Also, I ran these tests and I think I uncovered a bug: we don't canonicalize system jobs missing update block: https://github.com/hashicorp/nomad/blob/main/api/jobs.go#L1228-L1232
